linux/bindings/src/thread.rs

```diff
@@ -0,0 +1,271 @@
+                item_ptr as _,
+                crate::NUMA_NO_NODE,
+                Self::format_name(name).as_ptr() as _,
+            )
+        };
+        // SAFETY: FFI call.
+        if unsafe { kernel::bindings::IS_ERR(task_ptr as _) } {
+            // SAFETY: `item_ptr` is initialized above, and it should be dropped
+            // here due to the kthread creation failure.
+            unsafe { Box::from_raw(item_ptr) };
+            return Err(ENOMEM);
+        }
+        // SAFETY: it is checked above that `task_ptr` is valid.
+        Ok(unsafe { NonNull::new_unchecked(task_ptr) })
+    }
+
+    unsafe extern "C" fn threadfn<F, T>(data: *mut core::ffi::c_void) -> core::ffi::c_int
+    where
+        F: FnOnce() -> T,
+        F: Send,
+        T: Send,
+    {
+        // Execute the user closure.
+        let mut item = Box::from_raw(data as *mut ThreadItem<T>);
+        item.execute();
+        // Do not drop the item, until the kthread stops.
+        Box::into_raw(item);
+
+        return 0;
+    }
+
+    /// Returns true if the user specified `func` has been finished.
+    pub fn is_finished<T>(&self) -> bool {
+        // SAFETY: the `self.0` is a valid `task_struct`, so the `kthread_data`
+        // is also valid.
+        unsafe {
+            let item_ptr = crate::kthread_data(self.0.get()) as *mut ThreadItem<T>;
+            (*item_ptr).is_finished.load(Ordering::Acquire)
+        }
+    }
+
+    /// Blocks unless or until the current thread's token is made available.
+    pub fn park() {
+        // SAFETY: no safety requirements on this FFI call.
+        unsafe { crate::kthread_parkme() };
+    }
+
+    /// Makes the handle's token available if it is not already.
+    pub fn unpark(&self) {
+        // SAFETY: the `self.0` is a valid `task_struct`.
+        unsafe { crate::kthread_unpark(self.0.get()) };
+    }
+
+    // Cooperatively gives up a timeslice to scheduler.
+    pub fn yield_now() {
+        // SAFETY: no safety requirements on this FFI call.
+        unsafe { crate::schedule() };
+    }
+
+    /// Puts the current thread to sleep for at least the specified amount of time.
+    pub fn sleep(dur: Duration) {
+        // SAFETY: no safety requirements on this FFI call.
+        unsafe { crate::msleep(dur.as_millis() as _) };
+    }
+
+    /// Attempts to wake up the kthread.
+    pub fn wake_up(&self) {
+        // SAFETY: the `self.0` is a valid `task_struct`.
+        unsafe { crate::wake_up_process(self.0.get()) };
+    }
+
+    /// Stops the kthread and retrieves the data of `ThreadItem`.
+    ///
+    /// If the `threadfn` is never executed, the data is `None`, else it should
+    /// be some item produced by user specified closure.
+    ///
+    /// # Safety
+    ///
+    /// This method must only be called once, since we will drop `ThreadItem` here.
+    /// And after this, the `thread` should not be used ever.
+    pub unsafe fn stop<T>(&self) -> Option<T> {
```

Comment:
What if `wake_up` is called after a thread is stopped? Will any problem (e.g., undefined behaviors) be caused?